packages/dag/src/group.ts

```diff
@@ -191,22 +191,36 @@ const buildAcyclicGroupedGraphImpl = <T>(
   source: DataNodeMap<T>,
   groupKey: GroupKeyFunc,
   origGroupKey: GroupKeyFunc,
-): GroupDAG<T> => {
+  removedCycles: collections.set.SetId[][] = [],
+): { graph: GroupDAG<T>; removedCycles: collections.set.SetId[][] } => {
   // Build group graph
   const groupGraph = buildPossiblyCyclicGroupGraph(source, groupKey, origGroupKey)
   const possibleCycle = groupGraph.getCycle()
   if (possibleCycle === undefined) {
-    return new DAG(groupGraph.entries(), groupGraph.nodeData)
+    return { graph: new DAG(groupGraph.entries(), groupGraph.nodeData), removedCycles }
+  }
+  try {
+    const updatedGroupKey = modifyGroupKeyToRemoveCycle(groupGraph, groupKey, source, possibleCycle)
+    return buildAcyclicGroupedGraphImpl(source, updatedGroupKey, origGroupKey, removedCycles)
+  } catch (error) {
+    if (error instanceof CircularDependencyError) {
+      const { causingNodeIds } = error
+      log.debug(
+        'detected circular dependency in group graph, removing the following nodes: %s',
+        causingNodeIds.map(toString).join(', '),
+      )
+      causingNodeIds.forEach(nodeId => source.deleteNode(nodeId))
+      return buildAcyclicGroupedGraphImpl(source, groupKey, origGroupKey, removedCycles.concat([causingNodeIds]))
```

Comment:
IMO instead of passing `removedCycles` list into `buildAcyclicGroupedGraphImpl`, the outer call should return its `causingNodeIds`, with the `removedCycles` of the inner call:
```suggestion
      const { graph, removedCycles } = buildAcyclicGroupedGraphImpl(source, groupKey, origGroupKey)
      return { graph, removedCycles: [[causingNodeIds]].concat(removedCycles) }
```

Reply:
I found it harder to read this way 😅